### PR TITLE
PIM-7757: Fix concurent access on completeness table

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

 We have a lot of concurrent INSERT INTO to save the completeness in the database that generate SQL errors because of duplicated keys. To avoid the concurrent INSERT INTO that could generate a duplicated key, we register a Persister to Doctrine to change the SQL query generated by doctrine to replace the INSERT INTO by a REPLACE INTO only for the completeness objects.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Persister/CompletenessPersister.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Akeneo\Bundle\StorageUtilsBundle\Doctrine\Persister;
```
Comment (ahocquard) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Persister/CompletenessPersister.php]: As it concerns completeness of the product, it should maybe not be in storage utils but in the enrichment.

File: src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+use Akeneo\Component\StorageUtils\StorageEvents;
+use Doctrine\ORM\EntityManagerInterface;
+use Pim\Bundle\EnrichBundle\Doctrine\Persister\CompletenessPersister;
+use Pim\Component\Catalog\Model\Completeness;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+/**
+ *
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php]: useless empty lines to remove. Or better, to place a description in it :-)

File: src/Pim/Bundle/EnrichBundle/Resources/config/event_listeners.yml [hunk 3]
```diff
@@ -37,6 +37,13 @@ services:
         tags:
             - { name: kernel.event_subscriber }
 
+    pim_enrich.event_subscriber.perister_configurator:
+        class: Pim\Bundle\EnrichBundle\EventListener\PersisterConfiguratorSubscriber
+        arguments:
+            - '@pim_catalog.object_manager.product'
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Resources/config/event_listeners.yml]: '@doctrine.orm.entity_manager`, as the service you used is just an alias to it (probably a legacy from 1.x and MongoDB).

File: src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class PersisterConfiguratorSubscriber implements EventSubscriberInterface
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php]: Could you add a few phpspec on this, please?

File: src/Pim/Bundle/EnrichBundle/Doctrine/Persister/CompletenessPersister.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\Doctrine\Persister;
+
+use Doctrine\ORM\Persisters\Entity\BasicEntityPersister;
+
+/**
+ * We have a lot of concurrent INSERT INTO to save the completeness in the database that generate SQL errors because of
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Doctrine/Persister/CompletenessPersister.php]: :+1:

**Comment by damien-carcel** [comment]
Please update the changelog too.

File: src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php (hunk 4 above)
Comment (ahocquard) [reply src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php]: seriously, why are we specing everything? Don't mock what you don't own.

Specing the entity manager is totally useless. Same for the metadata.
What is the added value? Do you think you will dected any bug with a spec?

Integration tests for this kind of infra stuff.

Comment (damien-carcel) [reply src/Pim/Bundle/EnrichBundle/EventListener/PersisterConfiguratorSubscriber.php]: I can hear what you say. But it is the difference between specifications and unit tests. There is no need to unit test this, I agree, but phpspec is more than just about classic unit tests.
That being said, maybe we should only do unit tests using phpspec, and stop trying to write specifications for every classes.

**Comment by damien-carcel** [comment]
Closed in favor of https://github.com/akeneo/pim-community-dev/pull/9031

**Pull request closed without merging**